Add developer's docs on XBlock Asides (#1443)
openedx/docs.openedx.org: source/developers/references/developer_guide/extending_platform/xblock_asides.rst

context derived from ``block.scope_ids.usage_id``, or platform feature
flags.

Instance Methods
****************

``aside_view_declaration(view_name)``
=====================================

Return the bound method on this aside instance that is decorated with
``@XBlockAside.aside_for(view_name)``, or ``None`` if no such method
exists. The runtime uses this method to look up the aside's view function
when rendering. Subclasses do not normally need to call it directly.

``needs_serialization()``
=========================

Return ``True`` if the aside has any field whose value differs from its
default. The default implementation iterates over the aside's fields and
returns ``True`` if any field is set on this instance. The runtime calls
this method during OLX export to decide whether to serialize the aside as
an XML element. Subclasses rarely need to override this.

``parse_xml(node, runtime, keys)`` and ``add_xml_to_node(node)``
================================================================

Inherited from ``Blocklike``. Override these to customize how the aside is
read from and written to OLX. The default implementations serialize the

> aside's fields as XML attributes and child elements, identical to the

Reviewer: Note to self: verify this.

Runtime API
***********

The methods listed below are defined on
:class:`~xblock.runtime.Runtime` and govern Aside discovery, instantiation,
and rendering. Aside subclasses do not normally call these directly. They
are documented here so runtime implementors and Aside authors can
understand the lifecycle.

Discovery
=========

``runtime.applicable_aside_types(block)``
   Return the list of Aside type names that may apply to ``block``. The
   default implementation returns every Aside class registered through
   the ``xblock_asides.v1`` entry point. A runtime may override this to
   filter by user, course, or other context.

   The edx-platform LMS runtime overrides this through
   ``lms_applicable_aside_types`` in
   ``lms/djangoapps/lms_xblock/runtime.py``, which gates Aside rendering
   on the ``XBlockAsidesConfig`` Django configuration model. When the
   model's current revision has ``enabled=False``, no asides render in
   the LMS. When enabled, asides do not render on block types listed in
   the model's ``disabled_blocks`` field (default value:
   ``"about course_info static_tab"``). The Studio runtime does not
   apply this gate. See :ref:`Add an XBlock Aside` for the

>    administrative steps to enable the configuration.

Reviewer 2: As mentioned already, studio also gates the asides.

Author: @asadali145 it would be really helpful if you could suggest some ways to improve the docs here.

>    ``"about course_info static_tab"``). The Studio runtime does not
>    apply this gate. See :ref:`Add an XBlock Aside` for the
>    administrative steps to enable the configuration.

Reviewer 2: ```suggestion
   ``"about course_info static_tab"``). The Studio runtime gates it through 
   ``StudioConfig``. See :ref:`Add an XBlock Aside` for the
   administrative steps to enable the configuration.
```

   preserve client-side identification.

``runtime.wrap_aside(block, aside, view, frag, context)``
   Wrap an aside's fragment with a ``<div>`` carrying the aside's usage
   ID, the host block's usage ID, and any JavaScript initialization
   metadata. Override this if you need a different wrapping element or
   different ``data-`` attributes.

OLX Serialization
*****************

When a course is exported to OLX, the runtime serializes every aside
that returns ``True`` from ``needs_serialization()`` as an XML child
element of its host block. The XML tag of the aside element is the
aside's entry point name. Field values are written as attributes or
nested elements according to the aside's ``add_xml_to_node``
implementation.

On import, the runtime detects aside elements by looking up their tag
names in the registered ``xblock_asides.v1`` entry points. Tag names
that do not resolve to a registered aside are ignored. Field values are
then read by the aside's ``parse_xml`` implementation.

Two practical consequences:

* An aside's data only round-trips through OLX if both the source and
  destination platforms have the same aside installed under the same

>   entry point name.

Reviewer: Note to self: Verify this
